Android serialization mini-framework. #235

lemonboston opened the pull request:
Initial version submitted for discussion and improvements.

Commit:
Android serialization mini-framework. #235

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/core/Box.java:
```diff
@@ -0,0 +1,31 @@
+public interface Box<T> extends Parcelable
+{
+
+    T content();
```
As mentioned in Slack, please consider to add a `Context` parameter so we can potentially box stuff that requires a `Context` to be restored.

lemonboston replied:
(Note: I start to comment here, but will probably make updates for the other comments)

One thing that have come up now is that `Box<T>` can be used inside `Parcelable` implementations as well, and we don't have `Context` there in the `CREATOR`. 
Example:
```
    public void writeToParcel(Parcel dest, int flags)
    {
        dest.writeParcelable(mDelegate, flags);
        dest.writeParcelable(new DateTimeBox(mDate), flags);
    }

        public Dated createFromParcel(Parcel in)
        {
            ClassLoader classLoader = getClass().getClassLoader();
            return new Dated((
                    (DateTimeBox) in.readParcelable(classLoader)).content(),
                    (EventFilter) in.readParcelable(classLoader));
        }
```
(This is in current `serialization.temp.parcelable.Dated` - will write about those later..).

dmfs replied:
Hmm ... good point. I guess one solution would be to use the DateTimeBox internally instead of the date. Like so:

```
  private Box<DateTime> mDateBox;

  public Dated(DateTime date, EventFilter delegate)
  {
     this(new DateTimeBox(date), delegate);
  }

  private Dated(DateTimeBox datebox, EventFilter delegate)
  {
     mDateBox = datebox;
     mDelegate = delegate;
  }

  public void writeToParcel(Parcel dest, int flags)
    {
        dest.writeParcelable(mDelegate, flags);
        dest.writeParcelable(mDateBox, flags);
    }

        public Dated createFromParcel(Parcel in)
        {
            ClassLoader classLoader = getClass().getClassLoader();
            return new Dated(
                    (DateTimeBox) in.readParcelable(classLoader),
                    (EventFilter) in.readParcelable(classLoader));
        }
```

Also, how would that look if `Dated` doesn't implement `Box` but `Boxable` instead? Shouldn't that fix it too?

dmfs replied:
ah, scratch that. It would still not have a `Context`. So that's not an option. Ok I guess we can't do this then.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/SimpleKey.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * @author Gabor Keszthelyi
+ */
+public final class SimpleKey<V> implements Key<V>
```
How about the name `StringKey`? The key is derived from a `String` so that would make sense to me.

lemonboston replied:
Ok, changed it to `StringKey`.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/NestedBundle.java:
```diff
@@ -0,0 +1,47 @@
+            @Override
+            public Optional<Bundle> create()
+            {
+                return new NullSafe<>(intent.getBundleExtra(Keys.NESTED_BUNDLE));
```
How about wrapping this with `Obligatory` so the `NestedBundle` is always present (falling back to an empty bundle)?
```
return new Obligatory(new NullSafe<>(intent.getBundleExtra(Keys.NESTED_BUNDLE)), Bundle.EMPTY);
```
(optionally use `new Bundle()` instead of `Bundle.EMPTY` if the result should be mutable).

That way you wouldn't have to write
```
NestedBundle(getIntent()).value(null)
```
above. It could just be
```
NestedBundle(getIntent()).value()
```
If the `NestedBundle`, by contract, is always present.

lemonboston replied:
I thought about this `Obligatory` in the weekend and now I think that it can be misleading to use. I mean that wherever we need an `Obligatory<T>`, don't we actually need a `T`?
That's what its name means I think, so it's no longer an `Optional<T>`. If we pass it around as `Optional<T>` subsequent code will see it that way, providing its own defaults or routing on `isPresent()`, while it really should not at that point, because that is no longer a valid program flow.

As I see the motivation in this case, it would be to have a lazy `Bundle` that always evaluates to something. I think that's not the same as `Optional<Bundle>`.
Because `Bundle` is not an interface we can't do it in an implementation, so I guess the alternative to
```
Bundle bundle = new NestedBundle(getIntent()).value(null)
or
Bundle bundle = new NestedBundle(getIntent()).value(new Bundle());
```
would be a factory then..  but I am not sure that's worth it.

The contract that nested bundle would always present be - would than mean we also make the contract that all Fragments has a non-null argument?
The null check in `Argument` is because of that, so that `fragment.getArguments()` can be null, so just to be convenient to use `new Arugment(.. fragment.getArguments()). 
We could set up these contracts to eliminate nulls as much as possible, then we need to keep to that, making sure to add empty bundles in the parameterless Fragment factory methods.

Let me know if we should do this.

dmfs replied:
If you say it like that you could call every method a "conditional factory", but that's not the point.

Try to see it like this. An `Obligatory` like this:
```java
new Obligatory(optional, default);
```
is the declarative equivalent of these two expressions:
```java
optional.isPresent()? optional : new Present<>(default);

new Present(optional.value(default));
```
Also, note that `Obligatory` is a decorator to an existing `Optional`, it's not a type nor a "standalone class". I agree that there wouldn't be any point in an `Obligatory<T>` interface, because we could just use `<T>` instead. 

This `Optional` design is different from Java 8's `Optional` which is rather procedural in that you often use it to describe an algorithm ("if … is not present then fall back to …") whereas our `Optional` (and this `Obligatory` in particular) can be used to describe the *result* in a declarative way ("the value of …, falling back to … if necessary").

My idea is that we should not have to use `isPresent()` very often when we use optionals. For instance, if I pass an `Optional<Bundle>` to an `Extra`, the `Extra` should be absent if either the `Bundle` is absent or if it doesn't contain the extra. I don't want to call `isPresent` on the optional Bundle before I create the extra.

```java
  Optional<X> y = new Extra(new NestedBundle(intent));
```

Also it would be great if the same would work with a "builder":

```java
  Intent newIntent = new IntentBuilder("action").withExtra(key, new Extra(new NestedBundle(oldIntent))).build().
```
In that case `newIntent` should only contain the extra if it was present in `oldIntent`.

Or to give you a more practical example. `BaseActivity` contains this code:
```java
Extra<Integer> themeExtra = new Extra<>(Keys.THEME, getIntent());
if (themeExtra.isPresent())
{
      new IntentParameter<>(Keys.THEME, new IntBox(themeExtra.value())).into(intent);
}
return intent;
```

I think it should look like this:
```java
return new IntentBuilder(intent).withExtra(Keys.THEME, new Extra<>(Keys.THEME, getIntent())).build();
```
Resulting in an intent that contains the theme extra if and only if it was present in the original intent (i.e. build an Intent with whatever the current Intent contains as the theme)

So maybe also consider to update the builder adapter interface below with a method like this
```java
  // add an Optional extra only if present
  <T> BuilderAdapter withExtra(Key<T> key, Optional<T> value);
```
(names are subject to change).


So, to get back to the original issue: You don't have to apply `Obligatory` in `NestedBundle` but if you don't you should be more clear when you use it.
Your original code was 
```java
NestedBundle(getIntent()).value(null);
```
Which doesn't make any sense. Why would it default to `null`? Is it always expected to be present then just call `value()` (yes, we can't create a custom Bundle, that's a pitty, but even if we could, the interface would be unsuitable for this) if the `NestedBundle` is allowed to be absent, either use a proper fallback (if the method or constructor you were passing this to requires a value) or allow the constructor to take an `Optional` if absent values are valid parameters too.

lemonboston replied:
> new NestedBundle(getIntent()).value(null);

This was only replicating the previous usage (`getIntent().getExtras()`) where it could be null as well, but I agree it's better to always provide a Bundle arg for a Fragment which can have arguments.

>```
><T> BuilderAdapter withExtra(Key<T> key, Optional<T> value);
>```

We have `Key<T>` and `Box<T>` as parameters, we can't use general `<T>` or `Optional<T>` cause we don't know how to box it.

So that's why this won't work either:
> ```
> return new IntentBuilder(intent).withExtra(Keys.THEME, new Extra<>(Keys.THEME, getIntent())).build();
> ```

But I've add a `withCopiedExtra(Key<T> key, Intent intent)` to the builder, that works.

Btw, `NestedBundle` should really not be used normally, that's one aim of this design is to hide that boilerplate. It's internal to how we put extras into Intent. The usage in `EventListActivity` was only needed to take out the intent extras and forward as fragment bundle.

So I've added the initial version for the builders, please check them out.

I will write to you on Slack about the Obligatory and Optional usage topic.

dmfs replied:
Hmm. good point.
We could do that at least for `Boxable`:
```java
<T extends Boxable> BuilderAdapter withExtra(Key<T> key, Optional<T> value);
```

And for Optional Boxes
```java
<T> BuilderAdapter withExtra(Key<T> key, Optional<Box<T>> value);
```

That doesn't help here though. But I'm wondering if there isn't another way. Like a `BoxedExtra` that returns the boxed Object instead of the unboxed one. `BoxedExtra<T>` would be an `Optional<Box<T>>` and could be used with the method above. Or am I missing something?

I'm not a big fan of the `withCopiedParam` method. The use case looks way to limited to me. It doesn't work that well when I want to copy a Fragment Extra into an Intent or vice versa.

lemonboston replied:
Both of the above methods couldn't be added like this because they have the same erasure,
but I've added the second one and that `BoxedExtra<T>`. So it can work like that.
I've removed the `withCopiedParam`.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/Extra.java:
```diff
@@ -0,0 +1,40 @@
+            @Override
+            public Optional<T> create()
+            {
+                if (intent != null)
```
I guess this could be replaced by
```
return new Argument<T>(key, new NestedBundle(intent).value());
```

or even better, if `Argument` would take an `Optional<Bundle>` 
```
return new Argument<T>(key, new NestedBundle(intent)); 
```

dmfs replied:
Btw. don't check `intent` for `null`. If you pass a `null` Intent you're most likely doing it wrong and the code should not swallow this error silently.

lemonboston replied:
Right, simplified it and also removed the `null` check for `intent` - you're right that can never really be `null` normally.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/Argument.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class Argument<T> extends AbstractCachingOptional<T>
+{
+    // TODO Worth implementing here, not using the abstract class
```
Totally agree!

Also consider to accept `Optional<Bundle>` as a parameter ideally using lazy evaluation. So it could look like this:
```
   private final Key<T> mKey;
   private final Optional<Bundle> mBundle;
 
  public Argument(final Key<T> key, final Bundle bundle) 
  {
     this(key, new Present<>(bundle));
  }

  public Argument(final Key<T> key, final Optional<Bundle> bundle) 
  {
      mKey = key;
      mBundle = bundle;
  }

  boolean isPresent()
  {
     return mBundle.value(Bundle.EMPTY).contains(key.name());
  }

  // etc...
```

dmfs replied:
Also note, it should be safe to assume that the passed `Bundle` is never `null`. Otherwise the caller should wrap it with `NullSafe<>(bundle)`. Let's see if we can eliminate `null` checks altogether.

lemonboston replied:
It comes back to the same question, I think, whether we agree on a contract that all Fragments has to have an argument.
Just because of convenient usage:
```
// Convenient:
Event event = new Argument(Keys.EVENT, getArguments()).value();
// Non-convenient:
Event event = new Argument(Keys.EVENT, new NullSafe<>(getArguments())).value();
```
I think we should avoid having the need to write down the second one, so either we keep null check in `Argument` (either explicit or using `NullSafe`) or we introduce a contract to always have argument for Fragment.
Right?

lemonboston replied:
About implementing the `Optional`s directly instead of using the abstract class - I will rewrite all these classes later as it makes sense to optimize them since they will be core, often used classes.

lemonboston replied:
Oh, I just realized that a paremeterless Fragment will not have code to extract the argument! :)
Silly me. So we can safely assume that bundle is non-null when we actually want to use it.
I will update code accordingly.

lemonboston replied:
Done. Just ignore my previous comments on the `null` topic.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/boxes/DateTimeBox.java:
```diff
@@ -0,0 +1,80 @@
+    @Override
+    public void writeToParcel(Parcel dest, int flags)
+    {
+        dest.writeLong(mDateTime.getTimestamp());
```
you should also write time zone and maybe the all-day flag otherwise the reconstructed value might be incorrect.

lemonboston replied:
Ok, I've added them.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/eventlist/EventListFragment.java:
```diff
@@ -300,28 +299,17 @@ public void onDestroy()
-        {
-            return DateTime.nowAndHere();
-        }
+        return new Argument<>(Keys.DATE_TIME_START_AFTER, getArguments()).value(DateTime.nowAndHere());
```
It would be nice if we can find a simple mechanism to evaluate the fallback value lazily. There are cases in which the fallback value consumes many resources and it's a shame if we create it when it's never used. I just had this case somewhere else.
Something like the `Obligatory` decorator that takes a fallback factory instead of a specific fallback value could solve this.

lemonboston replied:
As mentioned above, I think we are talking about a factory here indeed, to create an object lazily based on conditions. Optional would be a misuse for this, I think.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/eventlist/EventListActivity.java:
```diff
@@ -41,7 +42,7 @@ protected void onCreate(Bundle savedInstanceState)
         {
             getSupportFragmentManager().beginTransaction().add(R.id.schedjoules_activity_content,
-                    EventListFragment.newInstance(getIntent().getExtras())).commit();
+                    EventListFragment.newInstance(new NestedBundle(getIntent()).value(null))).commit();
```
see my comment on `NestedBundle` below. The `null` fallback looks completely wrong to me.

lemonboston replied:
Discussed above.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/BasicFluentIntent.java:
```diff
@@ -0,0 +1,68 @@
+
+
+    @Override
+    public void into(Intent intent)
```
despite the fact that this `Parameter` interface looks procedural, this particular case could be solved more elegantly, without mutating an existing `Intent`.

I'm thinking of something like an `IntentBuilder`. Like so:
```
Intent newIntent = new IntentBuilder(oldIntent).withExtra(key, box).intent();
assertNotSame(oldIntent, newIntent);
```
or even like this (imitating the other constructors of an Intent):
```
Intent intent = new IntentBuilder(Intent.ACTION_VIEW, url).withExtra(key, box).intent();
```

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/core/Parameter.java:
```diff
@@ -0,0 +1,28 @@
+public interface Parameter<T>
+{
+
+    void into(T container);
```
I don't like this concept of mutating an existing container. Maybe consider using a builder style pattern like 

```
public interface Parametrized<T>
{
    T updated();

    Parametrized<T> withParameter(Key<V> key, V value);
}
```

Which would be used like this
```
Bundle newBundle = new Parametrized(oldBundle).withParameter(key, value).updated();
```

The builder should be immutable of course.

Not sure if that would work for all uses cases that you have in mind.

Also, the name `Parametrized` is not very good, because it would indicate a decorator, which is not the case. Maybe `BuildAdapter` would be more accurate (but it contains the pattern name, which is not very nice). I would call instances just like `BundleBuilder` or `IntentBuilder` omitting the "Adapter" part.

lemonboston replied:
I really like your idea to not mutate the container but return a modified copy! That was a basic problem here..
And I see there are copy constructors for both `Intent` and `Bundle`.

I am not sure though how well the above, combined `Parametrized<T>` would work..  
Wouldn't it work well only if we implement it once? For example, please check current `BundleParameter` and `ParcelableBundleParameter`. The 2nd one has input params `Key` and `Parcelable`, not `Key` and `Box`.
So I am wondering if it's better to keep them separate, the common implementations for the actual parameter entry and the builder.
So what do you think about this?:
```
interface Param<C>
{
    C updated(C originalCarrier);
}

interface FluentParam<C> extends Param<C>
{
    <T> FluentParam withParam(Key<T> key, Box<T> box);
}
```
Names can change of course. Maybe no need for the general `Fluent*`, we can have specific for `Intent`s with `withExtra()` and for `Bundles` with `withArg()`.

Usage could be this:
```
Bundle bundle = new BundleBuilder()
    .withParam(key1, box1)
    .withParam(key2, box2)
    .updated(new Bundle());

or for single param:
Bundle bundle = new BundleParam(key, box).updated(new Bundle());
Bundle bundle = new ParcelableBundleParam(key, parc).updated(new Bundle());
```

lemonboston replied:
Although, your usage looks better..  maybe it's better that way then, with one implementation and writing a bit more when we need to put the Box together.

lemonboston replied:
I will do it the way you suggested with dedicated interfaces for Intent and Bundle builder.

dmfs replied:
Not sure what you mean by "dedicated interfaces for Intent and Bundle builder.", but you probably can use the same interface for both of them. The only thing that would be different is the name of the actual build method I guess.

dmfs reviewed (changes requested):
done with the first round

lemonboston commented:
I've added examples of how to do the serialization for the `EventFilter` example,
so in general when
- we have our own type definition, and
- we will surely need to access internals for serialization, and
- we want decoration to still work

Four options for it:
1. Don't define serialization on the interface, just on the implementor classes -> isn't good because decorator couldn't serialize this way
2. The interface extends `Parcelable` -> this will be the good one, I think. We don't need `Box` at this point, doesn't add anything. We put it into a `Box` with `ParcelableBox` when we build the `Intent`/`Bundle`.
3. The interface implements `Box<T>` -> as mentioned above, no benefit over 2., but we have an extra method `content()` which always `return this`.
4. The interface implements `Boxable<T>` which has method `Box<T> boxed()` -> this looked like the best first, because it has some resemblance with what we did for `Equalable ListView.id()`, to make sure decoration works. But it seems this case is different. Both 3. and 4. just adds extra methods to the class (`content()`, `content()+boxed()`).

Small note: I learned that we can't implement `Parcelable` in an inner class, cause it can't have the creator static field. 

They are in the `serialization.temp` package, please check if you agree that we just need to go with
```
interface OurType extends Parcelable
```
in these cases.

dmfs commented:
You can make inner classes `Parcelable` if they are `static`. So you need to pass any state or optionally the object to parcel to the constructor. It's slightly more verbose but it's essentially what happens under the hood with a non static inner class.

This is the idea:
```java
public final class Dated implements EventFilter
{
   private final EventFilter mDelegate;
   private final DateTime mDateTime;

   / / constructors here

    @Override
    public Box<EventFilter> boxed()
    {
      return new DatedBox(this);
      // or probably better:
      return new DatedBox(mDelegate, mDateTime);
    }

  // other methods here …

  private final static class DatedBox implements Box<Dated>
  {
     private final Dated mDated;

     public DatedBox(Dated dated)
     {
        mDated = dated;
     }

   /*
     optional constructor to pass individual values instead of the Dated object.
     The advantage is that Java doesn't have to create "access(…)" methods

     With this method DatedBox doesn't even have to be a private inner class and
     could just be a regular public toplevel class.
  */
     public DatedBox(EventFilter filterDelegate, DateTime datetime)
     {
        mFilterDelegate = filterDelegate;
        mDateTime = datetime;
     }

     // all the parcelable stuff here access mDated methods like mDated.mDateTime
  }

}
```

dmfs commented:
Also note that in your case 4, the `EventFilter` interface wouldn't extend `Box` but just `Boxable`. So the only method added is `boxed()`. Whereas `Parcelable` adds 2 methods and a static field.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/temp/parcelable/Everything.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * @author Gabor Keszthelyi
+ */
+class Everything implements EventFilter
```
It just occurred to me that `AllEventsFilter` or `NoFilter` (depending on how you look at it) might be better names.

lemonboston replied:
Updated, although this is just temporary code for that here, I will remove them.

lemonboston commented:
I've added those variations on the `Boxable<T>` implementation. Although we need to implement the same things in all 3 main options, the fact that we can move it to another top level class with `Boxable<T>`, thus separating concerns does look better indeed. So 'real' logic is not hidden among  serialization boilerplate.
So I am on board now to use that one (`Dated_v3` + `DatedBox_v3`). Thanks.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/boxes/DateTimeBox.java:
```diff
@@ -0,0 +1,91 @@
+    @Override
+    public void writeToParcel(Parcel dest, int flags)
+    {
+        dest.writeString(mDateTime.getTimeZone().getID());
```
Note that the current `DateTime` implementation is quite old and will return `null` as the time zone for all-day events.

lemonboston replied:
Thanks, I've fixed it by simply passing null as `TimeZone` to the `DateTime` constructor in this case.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/BundleBuilder.java:
```diff
@@ -0,0 +1,75 @@
+    @Override
+    public <T> FluentBuilder<Bundle> withParam(Key<T> key, Optional<Box<T>> optBox)
+    {
+        Bundle bundle = new Bundle(mBundle);
```
if `optBox` is not present you can just return `this`, because the bundle is not going to be modified in that case.

lemonboston replied:
I was not sure about this, what would be the reasonable contract to this kind of immutable object 'with' method with optional parameter, should it return new instance for absent value or not.
But in practice, I see that we will probably not use the same builder to create multiple Bundles or Intents, so I changed it. I will also add note to the javadoc.

dmfs replied:
That's the beauty and the whole point of immutable objects. If we don't change anything we don't have to return a new instance and in fact there is no risk in returning the same instance and no benefit in returning a new instance. For the caller it doesn't matter at all.

Even if we use the same builder for multiple Bundles or Intents, this won't cause any problems, because the object is immutable. Any call to `with` that would change the state will create a new instance.

lemonboston replied:
You're right! :)

However `BundleBuilder` was not really immutable, I knew I would need to review them, I've done  it now and refactored a bit. So the mutable `Bundle` is copied now in both inbound and outbound.
`IntentBuilder` is changed at one point because of this.

(I will do one more full review of the branch shortly, check these ones, too.)

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/Extra.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * @author Gabor Keszthelyi
+ */
+public final class Extra<T> extends AbstractCachingOptional<T>
```
As you already mentioned, this should be called `OptionalExtra`

lemonboston replied:
Done.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/Extra.java:
```diff
@@ -0,0 +1,32 @@
+    // TODO Implement Optional here, not use abstract class
+
+
+    public Extra(final Key<T> key, final Intent intent)
```
Maybe consider creating a constructor that takes an activity or even context instead of an intent.

lemonboston replied:
Added the constructor with `Activity`.
(`Context` doesn't have generic way to retrieve its `Intent`, right?)

dmfs replied:
Right, a `Context` may not have an `Intent` (just think of the application context). I think only `Activity` and `Services` have `Intents`. So forget about the context idea ;-)

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/IntentBuilder.java:
```diff
@@ -0,0 +1,95 @@
+    @Override
+    public <T> FluentBuilder<Intent> withParam(Key<T> key, Optional<Box<T>> optBox)
+    {
+        FluentBuilder<Bundle> bundleBuilder = optBox.isPresent() ? mNestedBundleBuilder.withParam(key, optBox.value()) : mNestedBundleBuilder;
```
Same here, if `optBox` is not present just return `this`

lemonboston replied:
Updated this one as well.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/core/FluentBuilder.java:
```diff
@@ -0,0 +1,35 @@
+    <T> FluentBuilder<R> withParam(Key<T> key, Box<T> box);
+
+    <T> FluentBuilder<R> withParam(Key<T> key, Optional<Box<T>> optBox);
+}
```
I still think that having a method like
```
  <T extends Boxable> FluentBuilder<R> withParam(Key<T> key, Optional<T> box); 
```
could make things easier. You could give it another name to avoid signature conflicts.

lemonboston replied:
Added with `withOptionalBoxable` name (not the best possibly, let me know if you have better idea)

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/core/FluentBuilder.java:
```diff
@@ -0,0 +1,35 @@
+{
+    R build();
+
+    <T> FluentBuilder<R> withParam(Key<T> key, Box<T> box);
```
maybe consider to rename `withParam` to just `with`. the signature of the method makes pretty clear what it means.

lemonboston replied:
Good idea, thanks, I like these concise names here. (Updated)

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/core/FluentBuilder.java:
```diff
@@ -0,0 +1,38 @@
+
+    <T> FluentBuilder<R> with(Key<T> key, Box<T> box);
+
+    // TODO tell in javadoc that absent will not return new instance
```
That won't be necessary, because it really doesn't matter since all instances must be immutable anyway.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/IntentBuilder.java:
```diff
@@ -0,0 +1,107 @@
+    @Override
+    public <T extends Boxable> FluentBuilder<Intent> withOptionalBoxable(Key<T> key, Optional<T> box)
+    {
+        return with(key, box.isPresent() ? new Present<Box<T>>(box.value().boxed()) : Absent.<Box<T>>absent());
```
That looks a bit more complicated that it has to IMO.
Why do you call `with(Key<T> key, Optional<Box<T>> optBox)`?
how about calling `with(Key<T> key, Box<T> box)` like so?
```java
    if (box.isPresent())
    {
        return this;
     }
     return with(key, box.value().boxed());
```

lemonboston replied:
True, I've simplified them.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/BundleBuilder.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public <T extends Boxable> FluentBuilder<Bundle> withOptionalBoxable(Key<T> key, Optional<T> box)
+    {
+        return with(key, box.isPresent() ? new Present<Box<T>>(box.value().boxed()) : Absent.<Box<T>>absent());
```
see my comment in `IntentBuilder`

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/BoxedExtra.java:
```diff
@@ -0,0 +1,50 @@
+            @Override
+            public Optional<Box<T>> create()
+            {
+                Bundle nestedBundle = new NestedBundle(intent).value(Bundle.EMPTY);
```
I would remove these redundant variables and write a one-liner. I think having these variables doesn't improve readability.

lemonboston replied:
(This code is not present like this anymore)

lemonboston commented:
I've realized that we don't need separate classes for the parameter extraction from `Bundle` and `Intent`, just extra constructors. Added these general ones as `Parameter` and `OptionalParameter`.
By the way, I've also added the mandatory version as `Lazy`.

So I would suggest we keep only the general ones. The question is the name. And although I didn't like this `Argument` name for the `Bundle` version, if we keep only one, I would vote for that.
I've looked up once before what's the difference of 'parameter' and 'argument' and they say here for example:
http://stackoverflow.com/a/156778/4247460
that 'parameter' is more like the declaration, formal one and 'argument' is the actual value.
So for the extracting side I think 'argument' fits well, we have an actual value there to represent, not a general parameter.
So I suggest renaming current `Parameter` and `OptionalParameter` to `Argument` and `OptionalArgument` and remove the others.
What do you think?

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/BundleBuilder.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+
+    private BundleBuilder(Bundle bundle, String nonUsedParam)
```
I'd prefer a slightly different pattern to have this internal non-cloning constructor.
Consider this:

```
   public BundleBuilder() 
   {
       // we don't need to clone Bundle.EMPTY, because we don't try to mutate it
       this(Bundle.EMPTY, false);
   }

   public BundleBuilder(Bundle bundle) 
   {
       this(bundle, true);
   }

   private BundleBuilder(Bundle bundle, boolean clone)
   {
      mBundle = clone ? new Bundle(bundle) : bundle;
   } 
```

lemonboston replied:
You're right, that's nicer, thanks. I've updated it.

dmfs commented:
I've fine with "Argument"

lemonboston commented:
I've squashed and rebased and finalized everything, ready for review.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/utils/factory/Lazy.java:
```diff
@@ -26,6 +26,10 @@
+     * <p>
+     * Note:
+     * If <code>T</code> is mutable, the returned object is always the same instance.
+     * If <code>T</code> is immutable, the instance can be different (as there is no need for it to be the same).
```
These two notes are pointless. The important point is that the "identity" of the result is the same on every call. It doesn't matter if the result is mutable or not. I think the definition of `Lazy` already makes clear how it behaves. Besides, this method doesn't get any parameters, so the result is expected to be the same on every call (since the `Lazy` instance is supposed to be immutable).

lemonboston replied:
Ok, I removed them.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/utils/factory/AbstractLazy.java:
```diff
@@ -0,0 +1,41 @@
+
+    public AbstractLazy(Factory<T> factory)
+    {
+        mDelegate = new SimpleLazy<>(factory);
```
Why does `AbstractLazy` build on `SimpleLazy`? Wouldn't it make more sense to be the other way round?

lemonboston replied:
So true. I've changed it.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/core/FluentBuilder.java:
```diff
@@ -0,0 +1,62 @@
+     * Same as {@link #with(Key, Box)} but with taking an {@link Optional<Boxable>} to take the {@link Box} from.
+     * If that is absent, no change is applied.
+     */
+    <T extends Boxable> FluentBuilder<R> withOptionalBoxable(Key<T> key, Optional<T> optBoxable);
```
I think `withBoxable` should be sufficient for a name. `withOptionalBoxable` is certainly "more correct", but it's quite long and cumbersome. For consistency the other 2 methods could be called `withBox` then. I think the `Optional` doesn't have to be part of the name in this case, since it feels more like sort of an attribute rather than a type.
Does that make sense?

lemonboston replied:
Yes, I agree that `withBoxable` is sufficient, I've changed it.
I am not sure about `withBox` though, I think the method adds a parameter which is the key and the value together, so key is important as well. `withBox` would only refer to the value. I know it would be to be consistent with `withBoxable`, but I suspect `with` will be used much more often, so I'd like it to be that concise.
So as I see `withParameter`, `withParam` , `with` are appropriate names, `withBox` not so much.
So I would rather treat `withBoxable` as a necessary technical exception on this interface and keep the other two as `with`.
But let me know if you'd prefer them to be changed.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/OptionalArgument.java:
```diff
@@ -0,0 +1,95 @@
+
+    public OptionalArgument(Key<T> key, Fragment fragment)
+    {
+        this(key, fragment.getArguments());
```
This will cause an NPE if the fragment has no arguments `Bundle`.

lemonboston replied:
Right, good catch, thanks. Updated.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/NestedBundle.java:
```diff
@@ -0,0 +1,48 @@
+            public Bundle create()
+            {
+                Bundle nestedBundle = intent.getBundleExtra(Keys.NESTED_BUNDLE);
+                return nestedBundle == null ? new Bundle() : nestedBundle;
```
This class is a bit weird. It says "Represents the nested {@link Bundle} in an {@link Intent}.", first of all, it doesn't "represent" a `Bundle`, and secondly, what does it really represent if there is no nested Bundle in the intent?

lemonboston replied:
I've changed the description, please check if it makes more sense now.
I've also moved the nested bundle key to this class as it needs to be referenced by the general `IntentBuilder` as well, so it can't really be a project specific value.

Commit:
Remove unnecessary notes from Lazy javadoc.

Commit:
Change Lazy implementations: SimpleLazy now extends AbstractLazy, which contains the actual implementation.

Commit:
Change method withOptionalBoxable to withBoxable on FluentBuilder.

Commit:
Fix possible NPE in OptionalArgument constructor. Add that Fragment constructor to Argument as well.

Commit:
Change NestedBundle description, move the key to this class, as it needs to belong to the 'framework', not the project.

lemonboston commented:
Ready for check again.

dmfs reviewed (approved):
Ok, let's see it in action.
